Prevent memory leaks with inline MockMakers by using weak refs to mocks

moltmann opened the pull request:
Fixes #1313

This is needed for mock makers that store a map<weakref<mock>, invocation>. If the invocation holds a strong reference to the mock, the key never gets gc-ed and the map keeps on growing.

As weak references cannot be serialized, temporarily create a strong reference during serialization.

As the tests use invocations without keeping references to the mock we need to keep strong references in these cases.

mockitoguy commented:
I'll review this weekend. Thank you!

mockitoguy commented:
My understanding is that we need to update the code and use weak reference wrapper everywhere where we currently keep the mock objects. @moltmann diligently identified following spots: InterceptedInvocation.mock, RealMethodCall.instance, and SerializableRealMethodCall.instance. I reviewed the remaining code and found other instances safe (implementations of MockCreationListener)

@moltmann, I like your idea with SerializableReference! I'm assuming we need to use it also in RealMethodCall, right? I'd love to see a very descriptive exception message thrown when serializableReference.get() method returns null (rather than just returning null as it is the case for WeakReference.get().

Thank you for contribution!

moltmann commented:
@mockitoguy
- I added an error message to be thrown when SerializableReference returns null. This should never be seen for users of the public API. Internal users might run into this though.
- I think RealMethodCall does not need SerializableReference as it never gets serialized. I am not sure why I added it there.

mockitoguy commented:
Hey @moltmann, can you rebase on the current release/2.x? It contains your previous PR and it created some conflicts?

In the meantime, I'll complete the review of this so that we know what we need to do before we merge it.

mockitoguy commented:
@moltmann - great contribution, BIG THANKS!

I made tweaks on top of your changes. We're ready to merge once the build is green.

@moltmann, @raphw any final remarks before we merge it?

raphw commented:
LGTM!

moltmann commented:
lgtm2
